Update the virtual kubelet documentation (#1211)
liqotech/liqo: docs/pages/concepts/offloading/mutating-webhook.md

---

> title: Mutating webhook

Reviewer: I would be confused by this title. In the end, we're taking about setting the offloading strategy, so I would expect this as title. Then we can say that this is achieved through a mutating webhook, but this would be the implementation, not the main message in the title.

weight: 5
---

This section explains how the Liqo Webhook mutates Pods created within a Liqo-enabled namespace, enabling them to be possibly offloaded to remote clusters and enforcing user-specified constraints.

> As detailed in the [namespace offloading](/usage/namespace_offloading) section, the creation of a *NamespaceOffloading* resource inside a given namespace extends that namespace to remote clusters (possibly remapped, to prevent naming conflicts).

Reviewer: I would add something like:
The *NamespaceOffloading* resource is created on the local cluster.
(although the name "local" looks confusing to me, I would prefer the term "home", but anyway, this has to be coherent with the rest of the documentation, so feel free to use whatever looks better for you)

Additionally, each *NamespaceOffloading* resource states the offloading constraints specified by the user, and leveraged by the webhook to appropriately mutate the Pods in that namespace (i.e., adding the *toleration* to the virtual nodes *taint*, and modifying the *node affinities* field).
In particular:

1. The [PodOffloadingStrategy](/usage/namespace_offloading#selecting-the-pod-offloading-strategy) field, which determines at a high-level whether Pod offloading is enabled.
2. The [ClusterSelector](/usage/namespace_offloading#selecting-the-remote-clusters) field, which optionally selects a subset of virtual clusters for Pod offloading.

> ### Pod mutations

Reviewer: Is the title correct? We're talking about pod offloading strategies, so the title "pod mutations" looks misleading to me.

This section shows in detail the webhook behavior with each of the three possible `PodOffloadingStrategy` values, while considering a fixed `ClusterSelector` that targets all virtual nodes with label `liqo.io/region=us-west-1`:

```yaml
clusterSelector:
  nodeSelectorTerms:
  - matchExpressions:
    - key: liqo.io/region
      operator: In
      values:
      - us-west-1
```

#### LocalAndRemote PodOffloadingStrategy

The `LocalAndRemote` strategy allows Pods to be scheduled both on local nodes and a subnet of the virtual nodes.

> Hence, the webhook adds the *toleration* to the virtual nodes *taint*, as well as mutates the pod affinity introducing two (ORed) `NodeSelectorTerms`:

Reviewer: I felt hard to understand what ORed was.
I would suggest to replace the parenthesis with "(in logical OR)".

* The first `NodeSelectorTerm` enforces the constraints specified through the `ClusterSelector`, and applies to virtual nodes.
* The second `NodeSelectorTerm` enables Pod scheduling on all physical nodes, as they might not match the `ClusterSelector`.

Hence, the resulting `NodeAffinity` constraint applied to a Pod would resemble the following:

```yaml
nodeAffinity:
  requiredDuringSchedulingIgnoredDuringExecution:

>     nodeSelectorTerms:

Reviewer: Question because I'm ignorant.
How can we understand that the two matchExpressions are in logical OR? In the example below, the syntax looks very much the same, but the text says they are in logical AND.

      - matchExpressions:
        - key: liqo.io/region
          operator: In
          values:
          - us-west-1
      - matchExpressions:
        - key: liqo.io/type
          operator: NotIn
          values:
          - virtual-node
```

In case a given Pod already contains a set of `NodeAffinity` constraints, they are merged appropriately by the Liqo webhook with the ones above, ensuring that only the Nodes respecting the intersection of the restrictions can be selected as valid scheduling targets.

#### Local PodOffloadingStrategy

> The `Local` strategy does not allow Pods to be scheduled on virtual nodes, hence behaving from this point of view as if Liqo were not present (although other resources, such as services, might be reflected to enable remote consumption).

Reviewer: ```suggestion
The `Local` strategy does not allow Pods to be scheduled on virtual nodes, hence behaving from this point of view as if Liqo was not present (although other resources, such as services, might be reflected to enable remote consumption).
```

Consequently, in this case the webhook is not activated, and Pods are neither appended the *toleration* nor any additional *NodeAffinity* constraints.

#### Remote PodOffloadingStrategy

The `Remote` strategy forces Pods to be scheduled on virtual nodes only.

> Hence, the webhook adds the *toleration* to the virtual nodes *taint*, as well as mutates the pod affinity introducing a single `NodeSelectorTerm` with two (ANDed) `MatchExpressions`:

Reviewer: ANDed
==> in logical AND

* The first `MatchExpression` term enforces the constraints specified through the `ClusterSelector`.
* The second `MatchExpression` ensures Pods are scheduled on virtual nodes only (i.e., with the `liqo.io/type=virtual-node` label).

The resulting `NodeAffinity` constraint applied to a Pod resembles the following (in case it already contained a set of constraints, they would be merged accordingly):

```yaml
nodeAffinity:
  requiredDuringSchedulingIgnoredDuringExecution:
    nodeSelectorTerms:
      - matchExpressions:
        - key: liqo.io/region
          operator: In
          values:
          - us-west-1
        - key: liqo.io/type
          operator: In
          values:
          - virtual-node
```

> 

Reviewer: At the end of this section I would add a pointer to the proper liqoctl page that says how to set the above information through command line, just to help the reader to understand how this can be done (withouth having to manually modify the resources).

### Constraints violation

The `NamespaceOffloading` resource specifies a set of constraints that applies to all pods of that namespace.
Additionally, users can define additional and more specific requirements at the Pod level (i.e., through the `NodeAffinity` configuration), automatically merged by the Liqo webhook with the ones originating from the `NamespaceOffloading` resource.

> Still, care must be pay to prevent conflicting requirements, which would cause the Pod to remain in `Pending` status as unschedulable, as seen in the violation constraints example of the [Liqo Extended tutorial](/gettingstarted/extended/hard_constraints/).

Reviewer: ```suggestion
However, be careful about preventing conflicting requirements, which would cause the Pod to remain in `Pending` status as unschedulable, as seen in the violation constraints example of the [Liqo Extended tutorial](/gettingstarted/extended/hard_constraints/).
```
